### PR TITLE
Qute - improve support for type-safe fragments

### PR DESCRIPTION
**Pull request by mkouba**
This pull request follows up on https://github.com/quarkusio/quarkus/pull/28216.

### CODE REVIEW

File: docs/src/main/asciidoc/qute-reference.adoc [hunk 1]
```diff
@@ -1681,6 +1683,61 @@ public class ItemResource {
...
+
+@CheckedTemplate
+class Templates {
+
+  // defines a type-safe template
+  static native TemplateInstance items(List<Item> items);
+
+  // defines a fragment of Templates#items() with identifier "item"
+  @CheckedFragment
+  static native TemplateInstance items$item(Item item); <1>
```
Comment (gastaldi) [inline docs/src/main/asciidoc/qute-reference.adoc:1710]: So following the example below, this method should be named `items$item_li`?

**Review by gastaldi — APPROVED** [review]
Nice! Added a question for better understanding of this proposal

File: docs/src/main/asciidoc/qute-reference.adoc (hunk 1 above)
Comment (mkouba) [reply docs/src/main/asciidoc/qute-reference.adoc:1710]: Good catch! Yes, we should either change the method name to `items$item_li(Item item)` or change the fragment id to `item`.

**Comment by mkouba** [comment]
I'm going to merge this pull request. 

Devtools Tests - JDK 11 Windows fails with:
```
2022-10-12T06:27:43.4634688Z [ERROR] Failed to execute goal on project test-codestart: Could not resolve dependencies for project org.test:test-codestart:jar:1.0.0-codestart: Could not transfer artifact org.eclipse.microprofile.context-propagation:microprofile-context-propagation-api:jar:1.2 from/to central (https://repo.maven.apache.org/maven2): transfer failed for https://repo.maven.apache.org/maven2/org/eclipse/microprofile/context-propagation/microprofile-context-propagation-api/1.2/microprofile-context-propagation-api-1.2.jar: Connect to repo.maven.apache.org:443 [repo.maven.apache.org/146.75.28.215] failed: Connection timed out: connect 
```

And JDK 17 MacOS M1 seems to be broken these days.

**Comment by FroMage** [comment]
Guys, I'm really sorry that it takes me more than 3 days to review stuff, but while I like this feature, I wonder why we need a new annotation at all? IMO the presence of a `$` sign in the method is more than enough to indicate it's a fragment and the annotation is redundant and should be removed. WDYT?

Also, if we can define type-safe fragments, why not also figure out a way to define type-safe tags? Perhaps it's as simple as doing this:

```java
@CheckedTemplates
public class tags {
 public static native TemplateInstance user(User user);
}
```

for `tags/user.html`? This might even already be working?

**Comment by mkouba** [comment]
> Guys, I'm really sorry that it takes me more than 3 days to review stuff, but while I like this feature, I wonder why we need a new annotation at all? IMO the presence of a $ sign in the method is more than enough to indicate it's a fragment and the annotation is redundant and should be removed. WDYT?

No problem. I think that we need an annotation for (a) to make it safe because in theory a template with the `$` sign in name may already exist and (b) it's IMO more readable and more apparent.

> Also, if we can define type-safe fragments, why not also figure out a way to define type-safe tags? Perhaps it's as simple as doing this

Ofc, you can do type-safe tags and yes, `public class tags` should work but it's IMO better to follow the naming conventions and specify the `CheckedTemplate#basePath` like [this](https://github.com/mkouba/ddmwi/blob/main/src/main/java/com/github/mkouba/ddmwi/ctrl/Tags.java).

**Comment by FroMage** [comment]
> No problem. I think that we need an annotation for (a) to make it safe because in theory a template with the $ sign in name may already exist and (b) it's IMO more readable and more apparent.

Come on, nobody has templates with the `$` sign, because nobody uses that in file names, that's not a real argument. As to more apparent, I'd argue that a `$` sign anywhere in a method name SCREAMS something, and the annotation is just getting in the way.

> Ofc, you can do type-safe tags and yes, public class tags should work

Oh, cool, I didn't realise that. We should probably mention it explicitely in the docs?

**Comment by mkouba** [comment]
> Come on, nobody has templates with the $ sign, because nobody uses that in file names, that's not a real argument.

It's a valid argument. And it's always better to make it obvious: _"Why is the `$` sign in the name? Ah, I see there is the `@CheckedFragment` annotation so it's a fragment template."_

**Comment by FroMage** [comment]
That's too much complexity IMO, we should remove it, it's superfluous.

**Comment by mkouba** [comment]
> That's too much complexity IMO, we should remove it, it's superfluous.

-1

**Comment by FroMage** [comment]
OK, so we disagree, not the first time ;) How do we resolve this? Do I open a discussion, or a quarkus-dev email? To ask others for input, so we can figure out what others prefer.

**Comment by mkouba** [comment]
> OK, so we disagree, not the first time ;) How do we resolve this? Do I open a discussion, or a quarkus-dev email? To ask others for input, so we can figure out what others prefer.

Neither @sumitsarkar nor @gastaldi had objections when the API was discussed/reviewed in this pr and https://github.com/quarkusio/quarkus/pull/28216. Feel free to open a discussion if you believe it's worth it. I'm not so sure. It's funny that last week you did not think this feature could be even useful and now you're fighting for this change in the API ;-)

**Comment by sumitsarkar** [comment]
I am leaning more toward making it obvious with the annotation - hence no objection to this style. While I agree with the argument that generally nobody uses `$` in a filename, the filesystems allow adding `$` to a filename. So we'd come across incorrect behavior with this feature if it is not made explicit with the annotation.

**Pull request merged**